### ISSUE TITLE
Zonal Stats Issue

### ISSUE THREAD

**Issue by bangen** [Riverscapes/pyBRAT#155]
@banderson1618 In running some QA/QC I discovered that zonal stats isn't always calculating the correct value.  I detail the issue in this video [here](https://youtu.be/jfiKcnnKFxc).  Could you please spend some time researching/testing a solution to this issue?  If it looks like it's going to take a really long time to come up with a solution let me know and we can re-assess how to tackle the issue.

**Comment by banderson1618** [comment]
I've watched the video you just put together. That's a frustrating bug, good work tracking it down. I don't think it matters so much for things like the conflict values, but it could matter more for things like the minimum and maximum elevations, which would affect our slope, possibly significantly. We could alleviate that by pulling data directly from the slope raster. We generate it already, just so that we can make a layer off of it. We might as well use it, especially if it gives us better values.

Other than the slope, what else might be significantly affected by the ends of the buffer being chopped off? I can't think of anything other than slope that this significantly affects off the top of my head. If we do come up with a workaround that's not prohibitively slow, we can consider only applying it to attributes that would be badly affected.

Off the top of my head, there are some options that could work as workarounds.
- Running Zonal Stats for each buffer individually is theoretically possible, but computationally expensive in the worst way. I wouldn't be surprised if this increased the time taken to run BRAT Table by 10x, if not more. I don't think this is the way to go.

- We could use the [Extract Values to Points](http://desktop.arcgis.com/en/arcmap/10.3/tools/spatial-analyst-toolbox/extract-values-to-points.htm) tool to get values that don't need to use a buffer. Maybe we use midpoints? It would probably be slower, since we would have to write a shape file. There's a tool called "Extract Values to Table", but I'm not sure about whether I can get it to join back with the BRAT output network properly.

- We can also create a super small buffer, say 1 foot in radius, and use that in Zonal Stats. Because it's so small, it almost wouldn't overlap at all with the buffers around it, and it would still have all the data that's on the line (I assume). This would have the benefit of being a lot easier to implement than the first solution.

@bangen, what are your thoughts? If you have objections, questions, or new ideas, I'd love to hear them.

**Comment by bangen** [comment]
I'm adding @wally-mac to make sure he's aware of the issue and bring him in on the conversation.  I'm still not sure what the best solution is.

We're using different buffers to pull of different stats, so some of them are more likely affected than others:

- **Slope:** iGeo_Min and iGeo_Max - 30 m buffer at the reach start and end points
- **Drainage Area:** iGeo_DA - 100 m buffer at the reach midpoint
- **Vegetation:** 'iVeg_xxx' - 30 m or 100 m buffer of entire reach
- **Land Use:** 'iPC_LU' - 100 m buffer of entire reach
- **Conflict Distances:** Distance to roads, rails, canals - 30 m buffer of entire reach

I'm still concerned about the degree of overlap.  In the video I showed a reach where the degree of overlap isn't as significant some other areas.  Check out the figures below.  The top one shows that ReachID 53 that I highlighted in the video.  The bottom figures shows another area of the watershed.

![overlapbuffers_reach53](https://user-images.githubusercontent.com/6502954/44802644-237f4e80-ab71-11e8-8c62-b2162d498fe0.PNG)

![overlapbuffers](https://user-images.githubusercontent.com/6502954/44802652-29752f80-ab71-11e8-98c0-2d975a1049c8.PNG)

**Comment by wally-mac** [comment]
@banderson1618, I watched Sara's video and I think it's a good idea to try the idea that she brings up (zonal stats as a table and dictionary) if it ends up being too computationally expensive I suggest asking about this issue on GIS Stack Exchange https://gis.stackexchange.com/ to see if others have figured out a work around.

**Comment by banderson1618** [comment]
@wally-mac, not to disparage my past self, but running Zonal Stats on each buffer individually is similar to what I did to get the data for the tools I built on my own (the Grain Size, Hydrologic Regime, and Tributary Impact tools), and those tools were incredibly slow. Those tools are as slow to run as they are because I created a buffer and extracted data for each reach individually. Those tools took hours, sometimes a couple days to run, and they had far fewer variables to extract than the BRAT Table tool does. That's not a viable solution for BRAT. I've actually been planning on going back and changing those tools if there's ever any interest in running them again some time in the future.

@bangen, that package looks like it has potential. I'll look into it. Good find!
